File: games/football/evolution/ga.py
"""
Football v2 — genetic algorithm and numpy NN forward pass.

Responsibilities:
  - Represent brains as {weights: np.ndarray[1233], fitness: float, ...}
  - Hybrid fitness computation (goal diff + fallback anchor tournament)
  - Tournament selection (k=5 default)
  - Two-point crossover over flat weight arrays
  - Gaussian mutation with per-generation decay
  - Breeding: elitism + selection + crossover + mutation + random injection
  - Numpy-based NN forward pass for running brains in local_test.py

Fitness weights are hyperparameters — iterated against local_test.py until
top_fitness climbs measurably over 10 generations (phase 3 hard gate).

The numpy NN matches the JS nn.js architecture bit-for-bit: 20→20→16→18→9,
LeakyReLU hidden, tanh output, 1233 total parameters, flat weight layout
[weights0, biases0, weights1, biases1, ...]. Inputs 18 and 19 are
cos/sin of the player's heading.
"""
from __future__ import annotations

import math
import random
from dataclasses import dataclass
from typing import Optional

import numpy as np


# ── NN architecture (must match games/football/nn.js) ─────────

ARCH = [20, 20, 16, 18, 9]
LEAKY_SLOPE = 0.01


def _layer_offsets(arch: list[int]) -> list[int]:
    """Start offset of each layer in the flat weights array, plus the total."""
    offsets = [0]
    acc = 0
    for i in range(len(arch) - 1):
        fan_in = arch[i]
        fan_out = arch[i + 1]
        acc += fan_in * fan_out + fan_out
        offsets.append(acc)
    return offsets


LAYER_OFFSETS = _layer_offsets(ARCH)
WEIGHT_COUNT = LAYER_OFFSETS[-1]


# ── Fitness ────────────────────────────────────────────────────

@dataclass(frozen=True)
class FitnessWeights:
    """Tunable coefficients for the normalized [0, 1] fitness.

    `w_pop` and `w_fallback` must sum to 1 — the final fitness is guaranteed
    to stay in [0, 1] regardless of how the weights are split. `max_goal_diff`
    is the goal differential that saturates `pop_score` to 1.0 (e.g., 3 goals
    per match average means perfect pop dominance).
    """
    w_pop: float
    w_fallback: float
    max_goal_diff: float


def compute_fitness(brain: dict, w: FitnessWeights) -> float:
    """Normalized hybrid fitness, guaranteed in [0, 1].

    Formula:
        pop_score = clip(avg_goal_diff / max_goal_diff, -1, +1) mapped to [0, 1]
        fb_score  = (wins + 0.5 * draws) / fallback_matches     in [0, 1]
        fitness   = w_pop * pop_score + w_fallback * fb_score   in [0, 1]

    Semantics:
        0.0 = loses everything (or no data at all)
        0.5 = neutral (all draws, or pop_score average with no fallback data)
        1.0 = perfect dominance on both axes

    If a brain has no data on one axis, that axis contributes a neutral 0.5
    to the weighted sum so partial data still yields a meaningful score.
    A brain with no data at all returns 0.0.
    """
    pop_matches = brain.get("pop_matches", 0)
    fb_matches = brain.get("fallback_matches", 0)

    if pop_matches == 0 and fb_matches == 0:
        return 0.0

    if pop_matches > 0:
        avg_diff = brain.get("pop_goal_diff", 0.0) / pop_matches
        normalized = max(-1.0, min(1.0, avg_diff / w.max_goal_diff))
        pop_score = (normalized + 1) / 2
    else:
        pop_score = 0.5

    if fb_matches > 0:
        raw = (
            brain.get("fallback_wins", 0)
            + 0.5 * brain.get("fallback_draws", 0)
        ) / fb_matches
        # Defensive clamp in case wins + draws > matches due to bad data
        fb_score = max(0.0, min(1.0, raw))
    else:
        fb_score = 0.5

    return w.w_pop * pop_score + w.w_fallback * fb_score


# ── Tournament selection ───────────────────────────────────────

def tournament_select(population: list[dict], k: int, rng: random.Random) -> dict:
    """Pick k brains at random, return the one with the highest fitness.

    k=1 is random selection (no pressure), k=population_size is elitist
    (always returns the best). Default k=5 balances diversity and pressure.
    """
    sample = rng.sample(population, min(k, len(population)))
    return max(sample, key=lambda b: b["fitness"])


# ── Crossover ─────────────────────────────────────────────────

def two_point_crossover(
    parent_a: np.ndarray, parent_b: np.ndarray, rng: random.Random
) -> np.ndarray:
    """Two cut points partition the weight array into three segments. Child
    takes [outer, middle, outer] from [a, b, a] respectively.
    """
    n = len(parent_a)
    assert len(parent_b) == n, "parent weight counts must match"
    p1 = rng.randint(0, n - 1)
    p2 = rng.randint(0, n - 1)
    if p1 > p2:
        p1, p2 = p2, p1
    child = parent_a.copy()
    if p2 > p1:
        child[p1:p2] = parent_b[p1:p2]
    return child


# ── Mutation ──────────────────────────────────────────────────

def gaussian_mutate(
    weights: np.ndarray, rate: float, std: float, rng: random.Random
) -> np.ndarray:
    """Apply Gaussian noise to each weight with independent probability `rate`.
    Uses the provided random.Random for reproducibility (not numpy's global).
    """
    if rate <= 0.0:
        return weights.copy()
    out = weights.copy()
    # Use the random.Random so test seeding is deterministic. For speed,
    # generate masks/noises in numpy with a derived seed from rng.
    seed = rng.randrange(2**31)
    np_rng = np.random.default_rng(seed)
    mask = np_rng.random(len(weights)) < rate
    noise = np_rng.standard_normal(len(weights)) * std
    out[mask] += noise[mask]
    return out


# ── Weight init ────────────────────────────────────────────────

def he_init_weights(rng: random.Random) -> np.ndarray:
    """He initialization: N(0, sqrt(2/fan_in)) per weight matrix, biases 0.
    Uses the provided rng for determinism."""
    seed = rng.randrange(2**31)
    np_rng = np.random.default_rng(seed)
    weights = np.zeros(WEIGHT_COUNT, dtype=np.float64)
    idx = 0
    for i in range(len(ARCH) - 1):
        fan_in = ARCH[i]
        fan_out = ARCH[i + 1]
        stddev = math.sqrt(2 / fan_in)
        weights[idx : idx + fan_in * fan_out] = np_rng.standard_normal(fan_in * fan_out) * stddev
        idx += fan_in * fan_out
        idx += fan_out  # biases stay 0
    return weights


# ── Breeding ──────────────────────────────────────────────────

def breed_next_generation(
    population: list[dict],
    *,
    size: int,
    elitism: int,
    tournament_k: int,
    mutation_rate: float,
    mutation_std: float,
    random_injection_rate: float,
    rng: random.Random,
) -> list[dict]:
    """Produce the next generation from the current population.

    Steps:
      1. Sort by fitness descending; copy top `elitism` brains unchanged.
      2. Inject some fraction of fresh random brains (exploration).
      3. Fill the remainder by: select 2 parents via tournament_k, crossover,
         mutate, push into new population.

    New brains have reset fitness/match counters — only weights are copied.
    """
    sorted_pop = sorted(population, key=lambda b: b["fitness"], reverse=True)
    new_pop: list[dict] = []

    # 1. Elite pass-through
    for b in sorted_pop[:elitism]:
        new_pop.append(_fresh_brain(b["weights"].copy()))

    # 2. Random injection
    num_random = int(round(size * random_injection_rate))
    for _ in range(num_random):
        if len(new_pop) >= size:
            break
        new_pop.append(_fresh_brain(he_init_weights(rng)))

    # 3. Crossover + mutation fills the rest
    while len(new_pop) < size:
        parent_a = tournament_select(population, tournament_k, rng)
        parent_b = tournament_select(population, tournament_k, rng)
        child_weights = two_point_crossover(parent_a["weights"], parent_b["weights"], rng)
        child_weights = gaussian_mutate(child_weights, mutation_rate, mutation_std, rng)
        new_pop.append(_fresh_brain(child_weights))

    return new_pop[:size]


def _fresh_brain(weights: np.ndarray) -> dict:
    """Create a brain record with cleared fitness/match stats."""
    return {
        "weights": weights,
        "fitness": 0.0,
        "pop_matches": 0,
        "pop_goal_diff": 0.0,
        "blowout_bonus": 0.0,
        "fallback_matches": 0,
        "fallback_wins": 0,
        "fallback_draws": 0,
        "fallback_losses": 0,
    }


# ── Numpy NN forward pass ─────────────────────────────────────

def nn_forward(weights: np.ndarray, inputs: np.ndarray) -> np.ndarray:
    """Run the 20→20→16→18→9 network on one input vector.

    Layout matches nn.js: for each layer, [weights (fan_in * fan_out),
    biases (fan_out)] concatenated. LeakyReLU on hidden layers, tanh on output.
    """
    if len(weights) != WEIGHT_COUNT:
        raise ValueError(f"weight count mismatch: expected {WEIGHT_COUNT}, got {len(weights)}")
    if len(inputs) != ARCH[0]:
        raise ValueError(f"input size mismatch: expected {ARCH[0]}, got {len(inputs)}")

    current = np.asarray(inputs, dtype=np.float64)
    for layer in range(len(ARCH) - 1):
        fan_in = ARCH[layer]
        fan_out = ARCH[layer + 1]
        w_off = LAYER_OFFSETS[layer]
        b_off = w_off + fan_in * fan_out
        W = weights[w_off : b_off].reshape(fan_in, fan_out)
        b = weights[b_off : b_off + fan_out]
        z = current @ W + b
        is_output_layer = (layer == len(ARCH) - 2)
        if is_output_layer:
            current = np.tanh(z)
        else:
            current = np.where(z >= 0, z, z * LEAKY_SLOPE)
    return current

File: games/football/evolution/ga.mjs
/**
 * Football v2 — genetic algorithm.
 *
 * Pure module: no deps, no DOM, no filesystem. Consumed by broker.mjs and
 * warm-start tooling. Brain records use camelCase; the broker converts to
 * snake_case SQLite columns at the boundary.
 *
 * Determinism: all randomness flows through a seeded LCG (createGaRng)
 * matching the physics.js PRNG formula, so given the same seed the breeding
 * loop is bit-reproducible. The GA keeps its own rng stream independent of
 * physics so match rollouts don't perturb selection order.
 */

/* ── NN architecture (must match games/football/nn.js) ─────────── */

export const ARCH = [25, 16, 9];

const LAYER_COUNT = ARCH.length - 1;

function computeLayerOffsets(arch) {
  const offsets = [0];
  let acc = 0;
  for (let i = 0; i < arch.length - 1; i++) {
    acc += arch[i] * arch[i + 1] + arch[i + 1];
    offsets.push(acc);
  }
  return offsets;
}

export const LAYER_OFFSETS = computeLayerOffsets(ARCH);
export const WEIGHT_COUNT = LAYER_OFFSETS[LAYER_OFFSETS.length - 1];

/* ── Fitness ───────────────────────────────────────────────────── */

/**
 * Build a FitnessWeights record. `wPop` and `wFallback` should sum to 1 so
 * fitness stays in [0, 1]; `goalDiffScale` is the k in `tanh(avgDiff/k)`.
 * k≈2 goals makes a two-goal lead land at tanh(1)≈0.76 and a blowout
 * saturate smoothly rather than clip hard.
 */
export function makeFitnessWeights({ wPop, wFallback, goalDiffScale, maxGoalDiff }) {
  // Backwards-compat: if a caller still passes `maxGoalDiff`, use it as
  // the tanh scale. New callers should pass `goalDiffScale`.
  const k = goalDiffScale ?? maxGoalDiff ?? 2;
  return { wPop, wFallback, goalDiffScale: k };
}

/**
 * Normalized hybrid fitness in [0, 1]. A brain with zero data on both
 * axes scores 0.0; one axis of data yields neutral 0.5 on the missing
 * axis so partial data is still meaningful.
 *
 * Pop axis uses `tanh(avgGoalDiff / k)` rather than a hard clamp. That
 * choice (see 2026-04-17 brainstorm) is defined everywhere (no 0-0
 * divide-by-zero), smoothly saturates at blowouts, and gives a strong
 * signal on narrow wins without discarding information on big ones.
 */
export function computeFitness(brain, w) {
  const popMatches = brain.popMatches || 0;
  const fbMatches = brain.fallbackMatches || 0;

  if (popMatches === 0 && fbMatches === 0) return 0.0;

  let popScore;
  if (popMatches > 0) {
    const avgDiff = (brain.popGoalDiff || 0) / popMatches;
    const scale = w.goalDiffScale ?? w.maxGoalDiff ?? 2;
    // tanh(x) ∈ (-1, 1); shift to [0, 1].
    popScore = (Math.tanh(avgDiff / scale) + 1) / 2;
  } else {
    popScore = 0.5;
  }

  let fbScore;
  if (fbMatches > 0) {
    const raw =
      ((brain.fallbackWins || 0) + 0.5 * (brain.fallbackDraws || 0)) / fbMatches;
    fbScore = Math.max(0.0, Math.min(1.0, raw));
  } else {
    fbScore = 0.5;
  }

  return w.wPop * popScore + w.wFallback * fbScore;
}

/* ── Tournament selection ──────────────────────────────────────── */

/**
 * Fisher-Yates partial shuffle to draw `size` distinct indices from [0, n).
 * Deterministic given the rng.
 */
function sampleIndices(n, size, rng) {
  const pool = new Array(n);
  for (let i = 0; i < n; i++) pool[i] = i;
  const take = Math.min(size, n);
  for (let i = 0; i < take; i++) {
    // rng() is [0,1); floor(rng() * (n-i)) ∈ [0, n-i).
    const j = i + Math.floor(rng() * (n - i));
    const tmp = pool[i];
    pool[i] = pool[j];
    pool[j] = tmp;
  }
  return pool.slice(0, take);
}

export function tournamentSelect(population, k, rng) {
  const idx = sampleIndices(population.length, k, rng);
  let best = population[idx[0]];
  for (let i = 1; i < idx.length; i++) {
    const cand = population[idx[i]];
    if (cand.fitness > best.fitness) best = cand;
  }
  return best;
}

/* ── Crossover ─────────────────────────────────────────────────── */

/**
 * Two-point crossover over flat weight arrays. When the two cut points
 * collapse (p1 === p2) the child is a pure copy of parentA.
 */
export function twoPointCrossover(parentAWeights, parentBWeights, rng) {
  const n = parentAWeights.length;
  if (parentBWeights.length !== n) {
    throw new Error('parent weight counts must match');
  }
  let p1 = Math.floor(rng() * n);
  let p2 = Math.floor(rng() * n);
  if (p1 > p2) { const tmp = p1; p1 = p2; p2 = tmp; }
  const child = new Float64Array(n);
  child.set(parentAWeights);
  if (p2 > p1) {
    for (let i = p1; i < p2; i++) child[i] = parentBWeights[i];
  }
  return child;
}

/* ── Mutation ──────────────────────────────────────────────────── */

export function gaussianMutate(weights, rate, std, rng) {
  const out = new Float64Array(weights.length);
  out.set(weights);
  if (rate <= 0.0) return out;
  const sampler = gaussianFromRng(rng);
  for (let i = 0; i < out.length; i++) {
    if (rng() < rate) out[i] += sampler() * std;
  }
  return out;
}

/* ── Weight init ───────────────────────────────────────────────── */

/**
 * He initialization: N(0, sqrt(2/fan_in)) per weight matrix, biases left at
 * zero. Determinism flows entirely through the passed rng.
 */
export function heInitWeights(rng) {
  const weights = new Float64Array(WEIGHT_COUNT);
  const sampler = gaussianFromRng(rng);
  let idx = 0;
  for (let layer = 0; layer < LAYER_COUNT; layer++) {
    const fanIn = ARCH[layer];
    const fanOut = ARCH[layer + 1];
    const stddev = Math.sqrt(2 / fanIn);
    const matSize = fanIn * fanOut;
    for (let k = 0; k < matSize; k++) {
      weights[idx++] = sampler() * stddev;
    }
    idx += fanOut; // biases stay zero
  }
  return weights;
}

/* ── Breeding ──────────────────────────────────────────────────── */

/**
 * Produce the next generation.
 *   1. Sort current pop by fitness desc, copy top `elitism` unchanged.
 *   2. Inject round(size * randomInjectionRate) fresh random brains.
 *   3. Fill the remainder by tournament-select → crossover → mutate.
 *
 * New brains have cleared stats — only weights carry over from parents.
 */
export function breedNextGeneration(population, opts) {
  const {
    size,
    elitism,
    tournamentK,
    mutationRate,
    mutationStd,
    randomInjectionRate,
    rng,
  } = opts;

  const sorted = population.slice().sort((a, b) => b.fitness - a.fitness);
  const newPop = [];

  for (let i = 0; i < elitism && i < sorted.length; i++) {
    newPop.push(freshBrain(new Float64Array(sorted[i].weights)));
  }

  const numRandom = Math.round(size * randomInjectionRate);
  for (let i = 0; i < numRandom; i++) {
    if (newPop.length >= size) break;
    newPop.push(freshBrain(heInitWeights(rng)));
  }

  while (newPop.length < size) {
    const parentA = tournamentSelect(population, tournamentK, rng);
    const parentB = tournamentSelect(population, tournamentK, rng);
    let child = twoPointCrossover(parentA.weights, parentB.weights, rng);
    child = gaussianMutate(child, mutationRate, mutationStd, rng);
    newPop.push(freshBrain(child));
  }

  return newPop.slice(0, size);
}

/** Brain record with cleared fitness/match stats. */
export function freshBrain(weights) {
  return {
    weights,
    fitness: 0,
    popMatches: 0,
    popGoalDiff: 0,
    fallbackMatches: 0,
    fallbackWins: 0,
    fallbackDraws: 0,
  };
}

/* ── PRNG ──────────────────────────────────────────────────────── */

/**
 * Seeded LCG matching physics.js:createSeededRng() bit-for-bit. The GA keeps
 * its own stream so match rollouts don't perturb breeding order.
 */
export function createGaRng(seed) {
  let state = (seed >>> 0) || 1;
  return function rng() {
    state = (Math.imul(state, 1664525) + 1013904223) >>> 0;
    return state / 4294967296;
  };
}

/**
 * Box-Muller gaussian sampler built on a seeded rng. Returns a closure; each
 * call yields one standard normal sample (we discard the paired draw for
 * simplicity — the rng is cheap).
 */
export function gaussianFromRng(rng) {
  return function gauss() {
    const u1 = rng() || 1e-10;
    const u2 = rng();
    return Math.sqrt(-2 * Math.log(u1)) * Math.cos(2 * Math.PI * u2);
  };
}
